File: app/src/main/java/social/entourage/android/webview/WebViewFragment.java
package social.entourage.android.webview;


import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.support.v4.view.GestureDetectorCompat;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.webkit.SslErrorHandler;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import social.entourage.android.R;
import social.entourage.android.base.EntourageDialogFragment;

import static android.content.Context.CLIPBOARD_SERVICE;

/**
 *
 */

public class WebViewFragment extends EntourageDialogFragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = WebViewFragment.class.getSimpleName();

    private static final String REQUESTED_URL = "REQUESTED_URL";

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.webview_animated_layout)
    View animatedView;

    @BindView(R.id.webview_navigation_bar)
    View navigationView;

    @BindView(R.id.webview_title)
    TextView titleTextView;

    @BindView(R.id.webview_navigation_bar_menu)
    View menuView;

    @BindView(R.id.webview_navigation_bar_menu_background)
    View menuBackgroundView;

    @BindView(R.id.webview)
    WebView webView;

    @BindView(R.id.webview_progressbar)
    ProgressBar progressBar;

    private String requestedUrl;

    private GestureDetectorCompat gestureDetectorCompat;

    Animation bottomUpJumpAnimation;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public WebViewFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param requestedUrl Requested url as string.
     * @return A new instance of fragment WebViewFragment.
     */
    public static WebViewFragment newInstance(String requestedUrl) {
        WebViewFragment fragment = new WebViewFragment();
        Bundle args = new Bundle();
        args.putString(REQUESTED_URL, requestedUrl);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            requestedUrl = getArguments().getString(REQUESTED_URL);
        }
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        super.onCreateView(inflater, container, savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_webview, container, false);
        ButterKnife.bind(this, view);

        return view;
    }

    @Override
    public void onViewCreated(final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        showAnimation();
    }

    @Override
    protected @StyleRes int getSlideStyle() {
        return 0;
    }

    @Override
    public void dismiss() {
        webView.stopLoading();
        hideAnimation();
    }

    // ----------------------------------
    // Private methods
    // ----------------------------------

    private void showAnimation() {
        bottomUpJumpAnimation = AnimationUtils.loadAnimation(this.getContext(), R.anim.bottom_up);
        bottomUpJumpAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(final Animation animation) {}

            @Override
            public void onAnimationEnd(final Animation animation) {
                initialiseView();
            }

            @Override
            public void onAnimationRepeat(final Animation animation) {}
        });
        animatedView.startAnimation(bottomUpJumpAnimation);
    }

    private void hideAnimation() {
        Animation bottomDownJumpAnimation = AnimationUtils.loadAnimation(this.getContext(), R.anim.bottom_down);
        bottomDownJumpAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(final Animation animation) {}

            @Override
            public void onAnimationEnd(final Animation animation) {
                WebViewFragment.super.dismiss();
            }

            @Override
            public void onAnimationRepeat(final Animation animation) {}
        });
        animatedView.startAnimation(bottomDownJumpAnimation);
    }

    private void initialiseView() {
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebViewClient(new MyBrowser());
        webView.loadUrl(requestedUrl);

        // add a gesture detector to the navigation bar
        gestureDetectorCompat = new GestureDetectorCompat(this.getContext(), new NavigationViewGestureListener());
        navigationView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(final View v, final MotionEvent event) {
                boolean detectedUp = event.getAction() == MotionEvent.ACTION_UP;
                if (!gestureDetectorCompat.onTouchEvent(event) && detectedUp) {
                    return onUp(event);
                }
                return true;
            }
        });
    }

    private boolean onUp(final MotionEvent event) {
        animatedView.setTranslationY(0);
        return true;
    }

    // ----------------------------------
    // Click handling
    // ----------------------------------

    @OnClick(R.id.webview_back_button)
    protected void onBackClicked() {
        if (webView.canGoBack()) {
            webView.goBack();
        } else {
            dismiss();
        }
    }

    @OnClick(R.id.webview_more_button)
    protected void onMoreClicked() {
        toggleMenu();
    }

    @OnClick(R.id.webview_background)
    protected void onBackgroundClicked() {
        dismiss();
    }

    @OnClick(R.id.webview_menu_browser)
    protected void onMenuBrowserClicked() {
        Intent browseIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(webView.getUrl()));
        try {
            startActivity(browseIntent);
        } catch (Exception ex) {
            Toast.makeText(this.getContext(), R.string.no_browser_error, Toast.LENGTH_SHORT).show();
        }
        toggleMenu();
    }

    @OnClick(R.id.webview_menu_copy)
    protected void onMenuCopyClicked() {
        ClipboardManager clipboard = (ClipboardManager) getContext().getSystemService(CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText(webView.getUrl(), webView.getUrl());
        clipboard.setPrimaryClip(clip);
        Toast.makeText(getContext(), R.string.webview_copy_ok, Toast.LENGTH_SHORT).show();
        toggleMenu();
    }

    @OnClick(R.id.webview_menu_share)
    protected void onMenuShareClicked() {
        String shareString = getString(R.string.webview_share_text, webView.getUrl());
        Intent sharingIntent = new Intent(android.content.Intent.ACTION_SEND);
        sharingIntent.setType("text/plain");
        sharingIntent.putExtra(Intent.EXTRA_TEXT, shareString);
        startActivity(Intent.createChooser(sharingIntent, getString(R.string.entourage_share_intent_title)));
        toggleMenu();
    }

    @OnClick(R.id.webview_navigation_bar_menu_background)
    protected void toggleMenu() {
        menuView.setVisibility(menuView.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE);
        menuBackgroundView.setVisibility(menuBackgroundView.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE);
    }

    // ----------------------------------
    // MyBrowser
    // ----------------------------------

    private class MyBrowser extends WebViewClient {

        private String loadedUrl = "";

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            view.loadUrl(url);
            return true;
        }

        @Override
        public void onPageStarted(final WebView view, final String url, final Bitmap favicon) {
            super.onPageStarted(view, url, favicon);

            if (!loadedUrl.equalsIgnoreCase(url)) {
                loadedUrl = url;
                Uri uri = Uri.parse(url);
                String host = uri.getHost();
                String domain = "";
                if (host != null) {
                    host = host.toLowerCase();
                    if (host.startsWith("www.")) {
                        host = host.substring(4);
                    }
                    if (host.length() > 0) {
                        domain = host.substring(0, 1).toUpperCase() + host.substring(1);
                    }
                }
                titleTextView.setText(domain);

                progressBar.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public void onPageFinished(final WebView view, final String url) {
            super.onPageFinished(view, url);
            progressBar.setVisibility(View.GONE);
        }

        @Override
        public void onReceivedError(final WebView view, final WebResourceRequest request, final WebResourceError error) {
            super.onReceivedError(view, request, error);
            progressBar.setVisibility(View.GONE);
        }

        @Override
        public void onReceivedHttpError(final WebView view, final WebResourceRequest request, final WebResourceResponse errorResponse) {
            super.onReceivedHttpError(view, request, errorResponse);
            progressBar.setVisibility(View.GONE);
        }

        @Override
        public void onReceivedSslError(final WebView view, final SslErrorHandler handler, final SslError error) {
            super.onReceivedSslError(view, handler, error);
            progressBar.setVisibility(View.GONE);
        }
    }

    // ----------------------------------
    // Navigation gesture detector
    // ----------------------------------

    private class NavigationViewGestureListener extends GestureDetector.SimpleOnGestureListener {

        private boolean handleFling = true;

        @Override
        public boolean onDown(MotionEvent event) {
            handleFling = true;
            return true;
        }

        @Override
        public boolean onFling(MotionEvent event1, MotionEvent event2,
                               float velocityX, float velocityY) {
            // On fling down, dismiss the fragment
            if (event2.getRawY() - event1.getRawY() > 0 && velocityY < 0 && handleFling) {
                dismiss();
                return true;
            }
            return false;
        }

        @Override
        public boolean onScroll(final MotionEvent e1, final MotionEvent e2, final float distanceX, final float distanceY) {
            float translationY = animatedView.getTranslationY();
            float deltaY = e2.getRawY() - e1.getRawY();
            if (deltaY > 0) {
                animatedView.setTranslationY(deltaY);
            }
            if (translationY > deltaY) handleFling = false;
            return translationY > deltaY;
        }
    }

}

File: app/src/main/java/social/entourage/android/tools/view/WebViewFragment.kt
package social.entourage.android.tools.view

import android.annotation.TargetApi
import android.content.*
import android.content.pm.ResolveInfo
import android.graphics.Bitmap
import android.net.Uri
import android.net.http.SslError
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.webkit.*
import android.widget.Toast
import androidx.annotation.IdRes
import androidx.annotation.RequiresApi
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsIntent.SHARE_STATE_ON
import androidx.browser.customtabs.CustomTabsService.ACTION_CUSTOM_TABS_CONNECTION
import androidx.core.view.GestureDetectorCompat
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import social.entourage.android.EntourageApplication
import social.entourage.android.R
import social.entourage.android.base.BaseDialogFragment
import social.entourage.android.databinding.FragmentWebviewBinding
import timber.log.Timber
import java.lang.reflect.Executable
import java.util.*

class WebViewFragment : BaseDialogFragment() {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    private lateinit var binding: FragmentWebviewBinding
    private lateinit var requestedUrl: String
    @IdRes private var shareMessageRes: Int = 0
    private var gestureDetectorCompat: GestureDetectorCompat? = null
    var bottomUpJumpAnimation: Animation? = null

    private var hasToSendRead = false

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        super.onCreateView(inflater, container, savedInstanceState)
        binding = FragmentWebviewBinding.inflate(inflater, container, false)
        setHasOptionsMenu(true)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requestedUrl = arguments?.getString(REQUESTED_URL) ?: return dismiss()
        shareMessageRes = arguments?.getInt(SHARE_MESSAGE, 0) ?: 0
        hasToSendRead = arguments?.getBoolean(SENDGETREAD, false) ?: false
        showAnimation()
        binding.webviewBackButton.setOnClickListener {onBackClicked()}
        binding.webviewMoreButton.setOnClickListener {toggleMenu()}
        binding.webviewBackground.setOnClickListener {dismiss()}
        binding.webviewMenuBrowser.setOnClickListener {onMenuBrowserClicked()}
        binding.webviewMenuCopy.setOnClickListener {onMenuCopyClicked()}
        binding.webviewMenuShare.setOnClickListener {onMenuShareClicked()}
        binding.webviewNavigationBarMenuBackground.setOnClickListener {toggleMenu()}

        if (hasToSendRead) {
            setWebUrlRead()
        }
    }

    override val slideStyle: Int
        get() = 0

    override fun dismiss() {
        binding.webview?.stopLoading()
        hideAnimation()
    }

    private  fun setWebUrlRead() {
        EntourageApplication.get().apiModule.homeRequest
            .markRecoWebUrlRead(requestedUrl).enqueue(object : Callback<okhttp3.ResponseBody> {
                override fun onResponse(
                    call: Call<okhttp3.ResponseBody>,
                    response: Response<okhttp3.ResponseBody>
                ) {
                }

                override fun onFailure(call: Call<okhttp3.ResponseBody>, t: Throwable) {

                }
            })
    }
    private fun showAnimation() {
        bottomUpJumpAnimation = AnimationUtils.loadAnimation(this.context, R.anim.bottom_up)
        bottomUpJumpAnimation?.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                initialiseView()
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        binding.webviewAnimatedLayout?.startAnimation(bottomUpJumpAnimation)
    }

    private fun hideAnimation() {
        val bottomDownJumpAnimation = AnimationUtils.loadAnimation(this.context, R.anim.bottom_down)
        bottomDownJumpAnimation.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                super@WebViewFragment.dismiss()
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        binding.webviewAnimatedLayout?.startAnimation(bottomDownJumpAnimation)
    }

    private fun initialiseView() {
        binding.webview?.settings?.javaScriptEnabled = true
        binding.webview?.settings?.domStorageEnabled = true
        binding.webview?.webViewClient = MyBrowser()
        binding.webview?.loadUrl(requestedUrl)

        // add a gesture detector to the navigation bar
        this.context?.let { context ->
            gestureDetectorCompat = GestureDetectorCompat(context, NavigationViewGestureListener())
            binding.webviewNavigationBar.setOnTouchListener { _, event ->
                if (gestureDetectorCompat?.onTouchEvent(event) == true || event.action != MotionEvent.ACTION_UP) true
                else onUp(event)
            }
        }
    }

    private fun onUp(event: MotionEvent): Boolean {
        binding.webviewAnimatedLayout?.translationY = 0f
        return true
    }

    // ----------------------------------
    // Click handling
    // ----------------------------------
    private fun onBackClicked() {
        if (binding.webview?.canGoBack() == true) {
            binding.webview?.goBack()
        } else {
            dismiss()
        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        val _model = ViewModelProvider(requireActivity()).get(
            CommunicationRecoWebUrlHandlerViewModel::class.java)
        _model.setValid()
    }

    private fun onMenuBrowserClicked() {
        binding.webview?.url?.let {
            val browseIntent = Intent(Intent.ACTION_VIEW, Uri.parse(it))
            try {
                startActivity(browseIntent)
            } catch (ex: Exception) {
                Toast.makeText(this.context, R.string.no_browser_error, Toast.LENGTH_SHORT).show()
            }
            toggleMenu()
        }
    }

    private fun onMenuCopyClicked() {
        binding.webview?.url?.let {
            val clipboard = requireContext().getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText(it, it)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(context, R.string.webview_copy_ok, Toast.LENGTH_SHORT).show()
            toggleMenu()
        }
    }

    private fun onMenuShareClicked() {
        binding.webview?.url?.let { url ->
            val sharingIntent = getSharingIntent(requireContext(), url, shareMessageRes)
            context?.startActivity(Intent.createChooser(sharingIntent, getString(R.string.entourage_share_intent_title)))
            toggleMenu()
        }
    }

    private fun toggleMenu() {
        binding.webviewNavigationBarMenu.visibility = if (binding.webviewNavigationBarMenu.visibility == View.VISIBLE) View.GONE else View.VISIBLE
        binding.webviewNavigationBarMenuBackground.visibility = if (binding.webviewNavigationBarMenuBackground.visibility == View.VISIBLE) View.GONE else View.VISIBLE
    }

    // ----------------------------------
    // MyBrowser
    // ----------------------------------
    private inner class MyBrowser : WebViewClient() {
        private var loadedUrl = ""
        @Deprecated("Deprecated in Java")
        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            view.loadUrl(url)
            return true
        }

        //@ExperimentalStdlibApi
        override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
            super.onPageStarted(view, url, favicon)
            if (!loadedUrl.equals(url, ignoreCase = true)) {
                loadedUrl = url
                Uri.parse(url).host?.let { it->
                    var host = it.lowercase(Locale.ROOT)
                    if (host.startsWith("www.")) {
                        host = host.substring(4)
                    }
                    Log.wtf("wtf", "host: $host")
                    binding.webviewTitle.text = if (host.isNotEmpty()) host.replaceFirstChar {
                        if (it.isLowerCase()) it.titlecase(
                            Locale.ROOT
                        ) else it.toString()
                    } else ""

                } ?: run {
                    binding.webviewTitle.text = ""
                }
                binding.webviewProgressbar.visibility = View.VISIBLE
            }
        }

        override fun onPageFinished(view: WebView, url: String) {
            super.onPageFinished(view, url)
            binding.webviewProgressbar.visibility = View.GONE
        }

        @TargetApi(Build.VERSION_CODES.M)
        override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
            super.onReceivedError(view, request, error)
            Log.wtf("wtf", "error: ${error.description}")
            binding.webviewProgressbar?.visibility = View.GONE
        }

        //@SuppressWarnings("deprecation")
        @Deprecated("Deprecated in Java")
        override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
            super.onReceivedError(view, errorCode, description, failingUrl)
            binding.webviewProgressbar?.visibility = View.GONE
        }

        @RequiresApi(Build.VERSION_CODES.M)
        override fun onReceivedHttpError(view: WebView, request: WebResourceRequest, errorResponse: WebResourceResponse) {
            super.onReceivedHttpError(view, request, errorResponse)
            binding.webviewProgressbar?.visibility = View.GONE
        }

        override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
            super.onReceivedSslError(view, handler, error)
            binding.webviewProgressbar?.visibility = View.GONE
        }
    }

    // ----------------------------------
    // Navigation gesture detector
    // ----------------------------------
    private inner class NavigationViewGestureListener : SimpleOnGestureListener() {
        private var handleFling = true
        override fun onDown(event: MotionEvent): Boolean {
            handleFling = true
            return true
        }

        override fun onFling(event1: MotionEvent?, event2: MotionEvent,
                             velocityX: Float, velocityY: Float): Boolean {
            // On fling down, dismiss the fragment
            if (event2.rawY - (event1?.rawY ?: 0.0f) > 0 && velocityY < 0 && handleFling) {
                dismiss()
                return true
            }
            return false
        }

        override fun onScroll(e1: MotionEvent?, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
            val translationY = binding.webviewAnimatedLayout?.translationY ?: return false
            val deltaY = e2.rawY - (e1?.rawY ?: 0.0f)
            if (deltaY > 0) {
                binding.webviewAnimatedLayout?.translationY = deltaY
            }
            if (translationY > deltaY) handleFling = false
            return translationY > deltaY
        }
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        val TAG: String? = WebViewFragment::class.java.simpleName
        private const val REQUESTED_URL = "REQUESTED_URL"
        private const val SHARE_MESSAGE = "SHARE_MESSAGE"
        private const val SENDGETREAD = "sendRead"
        var customTabsPackages: ArrayList<ResolveInfo>? = null

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param requestedUrl Requested url as string.
         * @return A new instance of fragment WebViewFragment.
         */
        fun newInstance(requestedUrl: String?, @IdRes shareMessageRes: Int? = null, sendRead:Boolean): WebViewFragment {
            val fragment = WebViewFragment()
            val args = Bundle()
            args.putString(REQUESTED_URL, requestedUrl)
            args.putBoolean(SENDGETREAD,sendRead)
            shareMessageRes?.let {args.putInt(SHARE_MESSAGE, it) }
            fragment.arguments = args
            return fragment
        }

        /**
         * Returns a list of packages that support Custom Tabs.
         */
        fun getCustomTabsPackages(context: Context): ArrayList<ResolveInfo> {
            val pm = context.packageManager
            // Get default VIEW intent handler.
            val activityIntent = Intent()
                    .setAction(Intent.ACTION_VIEW)
                    .addCategory(Intent.CATEGORY_BROWSABLE)
                    .setData(Uri.fromParts("http", "", null))

            // Get all apps that can handle VIEW intents.
            val packagesSupportingCustomTabs: ArrayList<ResolveInfo> = ArrayList()
            pm.queryIntentActivities(activityIntent, 0).forEach { info ->
                val serviceIntent = Intent().apply {
                    this.action = ACTION_CUSTOM_TABS_CONNECTION
                    this.setPackage(info.activityInfo.packageName)
                }
                // Check if this package also resolves the Custom Tabs service.
                if (pm.resolveService(serviceIntent, 0) != null) {
                    packagesSupportingCustomTabs.add(info)
                }
            }
            return packagesSupportingCustomTabs
        }

        private fun getSharingIntent(context: Context, url: String, shareMessageRes: Int): Intent {
            val shareString = if(shareMessageRes!=0) context.resources.getString(shareMessageRes, url) else  url
            val sharingIntent = Intent(Intent.ACTION_SEND)
            sharingIntent.type = "text/plain"
            sharingIntent.putExtra(Intent.EXTRA_TEXT, shareString)
            return sharingIntent
        }

        fun launchURL(context: Context, url: String, @IdRes shareMessageRes: Int = 0): Boolean {
            if(customTabsPackages == null) customTabsPackages = getCustomTabsPackages(context)
            if(customTabsPackages.isNullOrEmpty()) return false
            //Create a PendingIntent to your BroadCastReceiver implementation

            val schemeParams = CustomTabColorSchemeParams.Builder()
                    .setToolbarColor(context.resources.getColor(R.color.accent))
                    .setSecondaryToolbarColor(context.resources.getColor(R.color.custom_button_accent_disabled))
                    .build()

            val customIntent = CustomTabsIntent.Builder()
                    .setStartAnimations(context, R.anim.bottom_up, R.anim.slide_out_to_right)
                    .setExitAnimations(context, R.anim.bottom_down, R.anim.slide_in_from_right)
                    .setDefaultColorSchemeParams(schemeParams)
                    .setShareState(SHARE_STATE_ON)
                    .build()

            try {
                customIntent.intent.putExtra(Intent.EXTRA_REFERRER, Uri.parse("android-app://" + context.packageName))
                customIntent.launchUrl(context, Uri.parse(url))
                return true
            }catch (e:Exception){
                //HERE Find true context


                return true
            }

        }
    }
}

class CommunicationRecoWebUrlHandlerViewModel : ViewModel() {
    val isValid:MutableLiveData<Boolean> = MutableLiveData()

    fun setValid() {
        isValid.value = true
    }
}